Add `RestartCount` to container metrics

keisku opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?

Add new metrics, `container.restarts`, which means the number of times a container has been restarted due to the restart policy.
For more details: https://docs.docker.com/engine/containers/start-containers-automatically/.

### Motivation

We have `kubernetes.containers.restarts` in Kubernetes environment, but not same metrics in container environment such as ECS.

### Describe how to test/QA your changes

Check if `container.restarts` is collected.

#### Docker

```bash
docker run -d --cgroupns host --pid host --name dd-agent \
-v /var/run/docker.sock:/var/run/docker.sock:ro \
-v /proc/:/host/proc/:ro \
-v /sys/fs/cgroup/:/host/sys/fs/cgroup:ro \
-e DD_API_KEY=xxx datadog/agent-dev:keisku-restart-count-py3

docker run --name startstop --restart always -d $(docker build -t startstop --quiet -<<-EOD
FROM busybox:latest
COPY --chmod=755 <<-EOF /start.sh
echo "Starting..."
sleep 5
echo "Exiting..."
exit 1
EOF
ENTRYPOINT /start.sh
EOD
)
```

<img width="1607" alt="2024-10-17_22-54-11" src="https://github.com/user-attachments/assets/562a18f1-b0e1-4aaa-a7f5-dd1176245e51">

#### Containerd

```bash
curl -L https://github.com/containerd/nerdctl/releases/download/v1.7.7/nerdctl-1.7.7-linux-amd64.tar.gz | sudo tar -C /usr/local/bin -xz nerdctl

curl -L https://github.com/moby/buildkit/releases/download/v0.16.0/buildkit-v0.16.0.linux-amd64.tar.gz | sudo tar -C /usr/local -xz

sudo buildkitd &

cat /tmp/ctx/Dockerfile
FROM busybox:latest
COPY --chmod=755 <<-EOF /start.sh
echo "Starting..."
sleep 5
echo "Exiting..."
exit 1
EOF
ENTRYPOINT /start.sh

sudo nerdctl run --name startstop --restart always -d $(sudo nerdctl build -t startstop --quiet /tmp/ctx/)

sudo nerdctl run -d --cgroupns host --pid host --name dd-agent \
-v /run/containerd/containerd.sock:/run/containerd/containerd.sock:ro \
-v /var/lib/cloud/data/instance-id:/var/lib/cloud/data/instance-id:ro \
-v /proc/:/host/proc/:ro \
-v /sys/fs/cgroup/:/host/sys/fs/cgroup:ro \
-e DD_API_KEY=xxxx \
-e DD_HOSTNAME_FILE=/var/lib/cloud/data/instance-id \
-e DD_LOG_LEVEL=debug \
-e DD_CRI_SOCKET_PATH=/run/containerd/containerd.sock \
datadog/agent-dev:keisku-restart-count-py3
```

<img width="1610" alt="2024-10-17_23-14-04" src="https://github.com/user-attachments/assets/a2505a16-aaac-4723-851c-4bb1fc5d8971">

#### ECS Fargate

- [Specifying a container restart policy in an Amazon ECS task definition](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/container-restart-policy-example.html)
- `DD_ECS_TASK_COLLECTION_ENABLED=true` for getting data from the task metadata endpoint v4

![2024-10-21_10-31-02](https://github.com/user-attachments/assets/2db6b10e-3121-4596-8622-db1e398fb1a2)

### Possible Drawbacks / Trade-offs

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

Commit:
add container.restart_count

Commit:
reno new

Commit:
fix licenses

clamoriniere commented on pkg/collector/corechecks/containers/containerd/adapter.go:
```diff
@@ -40,6 +40,7 @@ var metricsNameMapping = map[string]string{
 	"container.net.rcvd":                       "",                                         // Not present in legacy check
 	"containerd.blkio.service_recursive_bytes": "containerd.blkio.service_recursive_bytes", // Passthrough for custom metrics extension
 	"containerd.blkio.serviced_recursive":      "containerd.blkio.serviced_recursive",      // Passthrough for custom metrics extension
+	"container.restart_count":                  "containerd.restart_count",
```
we usually to add the type of the metrics in the name. 
can I suggest `container.restarts` instead like for `kubernetes.container.restarts`

Commit:
rename to container.restarts

Signed-off-by: keisku <keisuke.umegaki@datadoghq.com>

Commit:
improve release note

keisku commented:
When a container restarts due to the [restart policy](https://docs.docker.com/engine/containers/start-containers-automatically/#use-a-restart-policy), `container_id` remains unchanged.

```bash
docker run --name startstop --restart always -d $(docker build -t startstop --quiet -<<-EOD
FROM busybox:latest
COPY --chmod=755 <<-EOF /start.sh
echo "Starting..."
sleep 5
echo "Exiting..."
exit 1
EOF
ENTRYPOINT /start.sh
EOD
)
5ad6b491166faf611e84a82b65d3c82321251a71433ecca680efbfb023ec9d5d

docker inspect startstop | grep 'Id\|RestartCount'
        "Id": "5ad6b491166faf611e84a82b65d3c82321251a71433ecca680efbfb023ec9d5d",
        "RestartCount": 8,
```

neko-dd commented on releasenotes/notes/add-container-restart-count-b641319c2944395c.yaml:
```diff
@@ -8,4 +8,5 @@
 features:
   - |
-    Add ``container.restart_count`` with ``restart_policy`` tag.
+    Add new metrics, ``container.restarts``, which means the number of times a container has been restarted due to the restart policy.
```
```suggestion
    Add new metric, ``container.restarts``, which indicates the number of times a container has been restarted due to the restart policy.
```

keisku committed:
Update releasenotes/notes/add-container-restart-count-b641319c2944395c.yaml

Co-authored-by: Sandra (neko) <165049174+neko-dd@users.noreply.github.com>

Commit:
set restart_policy:unknown if it is unset

Signed-off-by: keisku <keisuke.umegaki@datadoghq.com>

Commit:
support container discovered by kubelet

Commit:
support container discoverd by ECS task metadata v4

Signed-off-by: keisku <keisuke.umegaki@datadoghq.com>

clamoriniere commented on comp/core/workloadmeta/collectors/internal/docker/docker.go:
```diff
@@ -297,6 +297,11 @@ func (c *collector) buildCollectorEvent(ctx context.Context, ev *docker.Containe
 			}
 		}
 
+		restartPolicy := "no"
```
it can be good to define the different restartPolicy value as a `const string` 
or even try to do a kind of golang enum

something like:
```golang
type RestartPolicyValueType string
const (
    RestartPolicyAlways RestartPolicyValueType = "always"
    RestartPolicyNo RestartPolicyValueType = "no"
)
```

Commit:
update cases where restartCount cannot be retrieved

Signed-off-by: keisku <keisuke.umegaki@datadoghq.com>

Commit:
define default restart policy

Signed-off-by: keisku <keisuke.umegaki@datadoghq.com>

mfpierre reviewed (approved):
LGTM for container-apps files

vboulineau commented on comp/core/workloadmeta/collectors/util/ecs_util.go:
```diff
@@ -122,6 +122,19 @@ func ParseV4TaskContainers(
+			restartCount = *container.RestartCount
+			// The restart policy in ECS is an object, not a string like in docker (no, always, on-failure).
+			// https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html#container_definition_restart_policy
+			restartPolicy = "ecs"
```
I am not sure why we end up with `ecs` while there are multiple possible values? Also is it the same container or a new container (i.e. same container id or not?)

vboulineau commented on pkg/collector/corechecks/containers/containerd/adapter.go:
```diff
@@ -40,6 +40,7 @@ var metricsNameMapping = map[string]string{
 	"container.net.rcvd":                       "",                                         // Not present in legacy check
 	"containerd.blkio.service_recursive_bytes": "containerd.blkio.service_recursive_bytes", // Passthrough for custom metrics extension
 	"containerd.blkio.serviced_recursive":      "containerd.blkio.serviced_recursive",      // Passthrough for custom metrics extension
+	"container.restarts":                       "containerd.restarts",
```
We don't have metrics to existing `containerd` and `docker` checks. Only the new `container` checks gets new metrics.

vboulineau commented on pkg/collector/corechecks/containers/docker/adapter.go:
```diff
@@ -38,8 +38,9 @@ var metricsNameMapping = map[string]string{
-	"container.net.rcvd":              "",                      // Removed, handled by custom network extension
+	"container.net.sent":              "", // Removed, handled by custom network extension
+	"container.net.rcvd":              "", // Removed, handled by custom network extension
+	"container.restarts":              "docker.restarts",
```
We don't have metrics to existing containerd and docker checks. Only the new container checks gets new metrics.

vboulineau commented on pkg/util/kubernetes/kubelet/types_kubelet.go:
```diff
@@ -57,6 +57,7 @@ type Spec struct {
 	SecurityContext   *PodSecurityContextSpec `json:"securityContext,omitempty"`
 	RuntimeClassName  *string                 `json:"runtimeClassName,omitempty"`
 	Tolerations       []Toleration            `json:"tolerations,omitempty"`
+	RestartPolicy     string                  `json:"restartPolicy,omitempty"`
```
Is it used anywhere, as the `container.restarts` metric does not make sense on Kubernetes.

vboulineau reviewed (changes requested):
I have some doubts about introducing this metric. The meaning is difficult to get right as it assumes the container will retain the same container ID, which is not true in Kubernetes and unclear on ECS.

Containerd being relatively rarely used directly (optionally on Nomad?), it leaves raw Docker, which I'm not sure it's worth due to the conflicting meaning on different platforms.

Commit:
Only the new container checks gets new metrics

Commit:
not add restart_policy tag

Signed-off-by: keisku <keisuke.umegaki@datadoghq.com>

keisku commented:
@vboulineau

Thanks for reviewing this PR! I addressed your comments.

> I have some doubts about introducing this metric. The meaning is difficult to get right as it assumes the container will retain the same container ID, which is not true in Kubernetes and unclear on ECS.

We have `kubernetes.containers.restarts` and this metric has same meaning as `container.restarts` in Kubernetes.

https://github.com/DataDog/datadog-agent/blob/7037764c5740960bf6d50265d9a7ca5e6bdccc18/pkg/collector/corechecks/containers/kubelet/provider/pod/provider.go#L155

The conditions under which a container restarts in Kubernetes are explained in the following documentation.

- [How Pods handle problems with containers](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#container-restarts)
- [Container restart policy](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy)

Both Kubernetes and ECS have the concept of `restartPolicy`, but in ECS, we have not been able to retrieve the restart count. Using `restartPolicy` in ECS offers the following benefits.

> You can enable a restart policy for each essential and non-essential container defined in your task definition, to overcome transient failures faster and maintain task availability. When you enable a restart policy for a container, Amazon ECS can restart the container if it exits, without needing to replace the task.
https://docs.aws.amazon.com/AmazonECS/latest/developerguide/container-restart-policy.html#:~:text=You%20can%20enable,replace%20the%20task.

keisku committed:
Merge branch 'main' into keisku/restart-count

Commit:
fix unit test: Only the new container checks gets new metrics

keisku commented:
/merge

vboulineau commented on pkg/collector/corechecks/containers/generic/processor.go:
```diff
@@ -192,6 +192,11 @@ func (p *Processor) processContainer(sender sender.Sender, tags []string, contai
 	}
 
+	restartCount := float64(container.RestartCount)
+	if 0 < restartCount {
```
nit: readability left->right

```suggestion
	if restartCount > 0 {
```

vboulineau commented on comp/core/workloadmeta/collectors/internal/kubelet/kubelet.go:
```diff
@@ -309,6 +309,7 @@ func (c *collector) parsePodContainers(
 				State:           containerState,
 				Owner:           parent,
 				Resources:       resources,
+				RestartCount:    container.RestartCount,
```
Should be removed as the `container.RestartCount` does not apply to this container id but to the container name

vboulineau commented on pkg/collector/corechecks/containers/generic/processor.go:
```diff
@@ -192,6 +192,11 @@ func (p *Processor) processContainer(sender sender.Sender, tags []string, contai
 		p.sendMetric(sender.Gauge, "container.pid.thread_limit", containerStats.PID.ThreadLimit, tags)
 	}
 
+	restartCount := float64(container.RestartCount)
```
nit: You don't need to cast before comparing, only if `> 0` and you can use `pointer.Ptr(float64(container.RestartCount))` directly

vboulineau reviewed (commented):
> We have kubernetes.containers.restarts and this metric has same meaning as container.restarts in Kubernetes.

It does not. The `kubernetes.containers.restarts` is not emitted with container-level tags (like container_id), only orchestrator level tag, because... containers are never restarted in Kubernetes, only new containers are created to replace previous one.

So currently your code on Kubernetes is incorrect. Basically the metric is useless on Kubernetes as it will always be 0 (or we'd need to have specific Kubernetes code in the generic `container` check, currently not worth it).

I verified that on ECS, the **same** container id is restarted, so the metric could bring value in that situation.

With that implemented, LGTM. Don't forget to update `integrations-core` to document the new metric.

Commit:
use pointer.Ptr()

Commit:
readability left->right

Commit:
remove RestartCount from kubelet collector

keisku commented:
@vboulineau Appreciate your review!

I've addressed all your code change requests (removed `RestartCount` on Kubernetes).
In my implementation, a different container ID is assigned for each `container.restarts`, so I am sure that `container.restarts` is different from `kubernetes.containers.restarts`. I had a misunderstanding. Thank you for pointing it out.

---

How I tested `container.restarts` on Kubernetes:

```bash
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Pod
metadata:
  name: restart-count-test
spec:
  containers:
    - name: startstop
      image: busybox
      command: ["sh", "-c", "sleep 60 && exit 1"]
EOF
```

<details><summary>kubectl get pod restartcount-test -o yaml</summary>

```bash
kubectl get pod restartcount-test -o yaml
.........
  containerStatuses:
  - containerID: containerd://39a66e6325f5a8fe45c66ea4d37fe2d30aa8b349072bc9048c3c4fd37428ff2c
    image: docker.io/library/busybox:latest
    imageID: docker.io/library/busybox@sha256:768e5c6f5cb6db0794eec98dc7a967f40631746c32232b78a3105fb946f3ab83
    lastState: {}
    name: startstop
    ready: false
    restartCount: 2
    started: false
    state:
      terminated:
        containerID: containerd://39a66e6325f5a8fe45c66ea4d37fe2d30aa8b349072bc9048c3c4fd37428ff2c
        exitCode: 1
        finishedAt: "2024-11-16T00:28:44Z"
        reason: Error
        startedAt: "2024-11-16T00:28:39Z"

# Wait 60 seconds and run again.
kubectl get pod restartcount-test -o yaml
  containerStatuses:
  - containerID: containerd://2ab252d94fd377508c5f276bd7b378d82d04ddbc8801c03d5f1e16b85211ca9a
    image: docker.io/library/busybox:latest
    imageID: docker.io/library/busybox@sha256:768e5c6f5cb6db0794eec98dc7a967f40631746c32232b78a3105fb946f3ab83
    lastState:
      terminated:
        containerID: containerd://2ab252d94fd377508c5f276bd7b378d82d04ddbc8801c03d5f1e16b85211ca9a # <--- Different container id
        exitCode: 1
        finishedAt: "2024-11-16T00:29:20Z"
        reason: Error
        startedAt: "2024-11-16T00:29:15Z"
    name: startstop
    ready: false
    restartCount: 3 # <--- Incremented
```

</details>

![2024-11-18_14-21-08](https://github.com/user-attachments/assets/9585fca0-4521-4cd3-a57e-afb7c71b9038)

`container.restarts` on Kubernetes had different `container_id` tag.

It seems that Kubelet increments the RestartCount by 1 before starting a container during its recreation.

https://github.com/kubernetes/kubernetes/blob/bf8354d460263e7c3518282010e6a96a63c7dafb/pkg/kubelet/kuberuntime/kuberuntime_container.go#L217-L222

`container.restarts` with `container_id` dispalys unexpected values in the Metrics Explorer when using space aggregation methods like avg or sum. For this reason, I believe `container.restarts` on k8s is unnecessary and have removed it from the PR.

keisku commented:
/merge

keisku committed:
Merge branch 'main' into keisku/restart-count

keisku commented:
/merge
